### ISSUE TITLE
incorrect conversion for half-hour time zones in western hemisphere

### ISSUE THREAD

**Issue by ostueker** [felipec/git-remote-hg#48]
git-remote-hg miscalculates the time-zone offsets for time zones that don't have a full-hour offset and are located west of UTC (e.g. St. John's, Newfoundland).

Basically it's caused because `int(3.5) == 3` but `int(-3.5) != 3`.

The attached script demonstrates the issue:

```
$ python convert_timezone.py 

Implementation in git-remote-hg (fails for St. John's)
------------------------------------------------------
Name         | ISO TZ  int tz  | gittz(tz)  pytz(TZ)   | iso->int->iso 
Tehran       |  +0330  -12600  |   +0330     -12600    |   +0330   
Berlin       |  +0100   -3600  |   +0100      -3600    |   +0100   
UTC          |  +0000     0    |   +0000        0      |   +0000   
St. John's   |  -0330   12600  |   -0430  *   10200  * |   -0310  *
Halifax      |  -0400   14400  |   -0400      14400    |   -0400   
Toronto      |  -0500   18000  |   -0500      18000    |   -0500   
Calgary      |  -0700   25200  |   -0700      25200    |   -0700   
Vancouver    |  -0800   28800  |   -0800      28800    |   -0800   

Correct implementation
----------------------
Name         | ISO TZ  int tz  | gittz(tz)  pytz(TZ)   | iso->int->iso 
Tehran       |  +0330  -12600  |   +0330     -12600    |   +0330   
Berlin       |  +0100   -3600  |   +0100      -3600    |   +0100   
UTC          |  +0000     0    |   +0000        0      |   +0000   
St. John's   |  -0330   12600  |   -0330      12600    |   -0330   
Halifax      |  -0400   14400  |   -0400      14400    |   -0400   
Toronto      |  -0500   18000  |   -0500      18000    |   -0500   
Calgary      |  -0700   25200  |   -0700      25200    |   -0700   
Vancouver    |  -0800   28800  |   -0800      28800    |   -0800   

$ python convert_timezone.py test
**********************************************************************
File "convert_timezone.py", line 12, in __main__.gittz
Failed example:
    gittz(int(3.5*3600))    # St. John's (fails!)
Expected:
    '-0330'
Got:
    '-0430'
**********************************************************************
File "convert_timezone.py", line 47, in __main__.pytz
Failed example:
    pytz('-0330')   # St. John's (fails!)
Expected:
    12600
Got:
    10200
**********************************************************************
2 items had failures:
   1 of   4 in __main__.gittz
   1 of   4 in __main__.pytz
***Test Failed*** 2 failures.
```

Pull-request with fix is coming up.
[convert_timezone.txt](https://github.com/felipec/git-remote-hg/files/54584/convert_timezone.txt)

**Comment by fingolfin** [comment]
Also fixed in my git-remote-hg fork, see https://github.com/fingolfin/git-remote-hg

**Comment by felipec** [comment]
Thanks for taking the time to explain in great detail the issue. The code is indeed wrong and I've applied your patches with some modifications.

I cleaned up your test (0dfae24d218a7780ce456f59cc2b6f2224bcc753), to make it simpler and more in line with the other rests. Also, keep in mind that there's no need to `cd..` inside a subshell; the environment is restored after it.

I also took the liberty to modify the commit message to add an explanation of what's the issue (e17e147fb165b5c635e45c07a9aeceaa9db7fc7a). However, you said `int(-3.5) != 3`, but it is. The issue is that `-33 // 10 != 3`.

Also, I opted for slightly more verbose code in order to help readability, but I checked that it works both on your test script, and the tests we both wrote for the test suite:

```python
def gittz(tz):
    sign = 1 if tz >= 0 else -1
    hours, minutes = divmod(abs(tz), 60 * 60)
    return '%+03d%02d' % (-sign * hours, minutes / 60)

def hgtz(tz):
    tz = int(tz)
    sign = 1 if tz >= 0 else -1
    hours, minutes = divmod(abs(tz), 100)
    tz = hours * 60 * 60 + minutes * 60
    return -sign * tz
```

Sorry for the long delay, but it's pushed now. Thanks!
